### ISSUE TITLE
finalize heartbeat feature

### ISSUE THREAD

**Issue by biblicabeebli** [onnela-lab/beiwe-backend#375]
### To Build

We will be pushing this feature to the main branch in the first week of June.

There is an ongoing technical discussion of how "useful" this is in terms of improving persistence for the Android app, _but we know it doesn't hurt._

### Finished details
- Heartbeat needs to have a study-configurable message.
- Needs to have a configurable duration of inactivity before triggering.
  - we will use a single configurable value that also covers the periodicity of (time between successive) notifications.
- ~need to test behavior of receiving an empty messages on ios~
  - iOS always hits the download surveys endpoint even when there are no survey ids in the notification.
  - Confirmed expected behavior after fixing some unrelelated bugs for ios version 2.5.1.
- Update the notifications page to include heartbeats.
- Need to test behavior of receiving an empty messages on android.
  - It just ignores it - android app receipt of messages has been built, see https://github.com/onnela-lab/beiwe-android/issues/88
- Currently logic is sending these notifications out with a 6 minute offset like 1:00 -> 1:06 due to some detail of the logic interacting with the 6 minute periodicity, fix that.
  - done.
- Need to add a minimum periodicity detail. Fundamental limit is 6, 30 minutes as the lowest dropdown option on the website study settings page seems good enough? but also 0 needs to work...
  - there is now a minimum value of 30...

### Other decisions:
- We are going to leave the uniqueness nonce as presently configured. this is literally a feature for nagging study participants to reopen the app, we don't care if there are problems caused by a lot of notifications stacking up.
- ~needs to be toggleable (per study)~
  - ... wait why would we ever need or want to disable this feature for a whole study?
  - ~we will stick a value of 0 for the time limit as disabled.~
- ~Will need to be ios-only for now - https://github.com/onnela-lab/beiwe-android/issues/88~
- ~This feature going live is going to be blocked by https://github.com/onnela-lab/beiwe-ios/issues/47 because part of that issue is checking what the behavior of multiple stacked up notifications does when they are received by the app.~
  - Nope, app handled that case correctly.

**Comment by biblicabeebli** [comment]
@hydawo I can't think of any administrative situation under which we would want to disable this feature. Can you?

**Comment by biblicabeebli** [comment]
Ok configurable heartbeat has been implemented, need to shift to iOS and then Android before we can remove the Experiment requirement.

**Comment by biblicabeebli** [comment]
Woops need to add minimum periodicity.  Fundamental limit is 6, but ... 30 minutes to save our users from themselves....

**Comment by biblicabeebli** [comment]
need to test behavior of empty receiving an empty messages on ios and android...

**Comment by biblicabeebli** [comment]
update the notifications page to include heartbeats

**Comment by hydawo** [comment]
`needs to be toggleable (per study)`
I don't think we need a toggle since studies can set the periodicity to a high value if they really want the featured "turned off" (mostly)

**Comment by biblicabeebli** [comment]
there is now a commit on the finish-heartbeat branch that removes the experiment field and consequently globally enables the feature.

**Comment by biblicabeebli** [comment]
moving this issue to done on our internal tracker, will close when it is on main branch.

**Comment by biblicabeebli** [comment]
this has been merged into the main branch.
